lib/evaluate/intrinsics.cc

```diff
@@ -0,0 +1,1244 @@
+static constexpr TypePattern KINDComplex{Complex, KindCode::effectiveKind};
+static constexpr TypePattern KINDChar{Char, KindCode::effectiveKind};
+static constexpr TypePattern KINDLogical{Logical, KindCode::effectiveKind};
+
+// The default rank pattern for dummy arguments and function results is
+// "elemental".
+ENUM_CLASS(Rank,
+    elemental,  // scalar, or array that conforms with other array arguments
+    elementalOrBOZ,  // elemental, or typeless BOZ literal scalar
+    scalar, vector,
+    shape,  // INTEGER vector of known length and no negative element
+    matrix,
+    array,  // not scalar, rank is known and greater than zero
+    known,  // rank is known and can be scalar
+    anyOrAssumedRank,  // rank can be unknown
+    conformable,  // scalar, or array of same rank & shape as "array" argument
+    reduceOperation,  // a pure function with constraints for REDUCE
+    dimReduced,  // scalar if no DIM= argument, else rank(array)-1
+    dimRemoved,  // scalar, or rank(array)-1
+    rankPlus1,  // rank(known)+1
+    shaped,  // rank is length of SHAPE vector
+)
+
+ENUM_CLASS(Optionality, required, optional,
+    defaultsToSameKind,  // for MatchingDefaultKIND
+    defaultsToDefaultForResult,  // for DefaultingKIND
+    repeats,  // for MAX/MIN and their several variants
+)
+
+struct IntrinsicDummyArgument {
+  const char *keyword{nullptr};
+  TypePattern typePattern;
+  Rank rank{Rank::elemental};
+  Optionality optionality{Optionality::required};
+  std::ostream &Dump(std::ostream &) const;
+};
+
+// constexpr abbreviations for popular arguments:
+// DefaultingKIND is a KIND= argument whose default value is the appropriate
+// KIND(0), KIND(0.0), KIND(''), &c. value for the function result.
+static constexpr IntrinsicDummyArgument DefaultingKIND{"kind",
+    {Int, KindCode::kindArg}, Rank::scalar,
+    Optionality::defaultsToDefaultForResult};
+// MatchingDefaultKIND is a KIND= argument whose default value is the
+// kind of any "Same" function argument (viz., the one whose kind pattern is
+// "same").
+static constexpr IntrinsicDummyArgument MatchingDefaultKIND{"kind",
+    {Int, KindCode::kindArg}, Rank::scalar, Optionality::defaultsToSameKind};
+static constexpr IntrinsicDummyArgument OptionalDIM{
+    "dim", {Int, KindCode::dimArg}, Rank::scalar, Optionality::optional};
+static constexpr IntrinsicDummyArgument OptionalMASK{
+    "mask", AnyLogical, Rank::conformable, Optionality::optional};
+
+struct IntrinsicInterface {
+  static constexpr int maxArguments{7};
+  const char *name{nullptr};
+  IntrinsicDummyArgument dummy[maxArguments];
+  TypePattern result;
+  Rank rank{Rank::elemental};
+  std::optional<SpecificIntrinsic> Match(const CallCharacteristics &,
+      const IntrinsicTypeDefaultKinds &,
+      parser::ContextualMessages &messages) const;
+  std::ostream &Dump(std::ostream &) const;
+};
+
+static const IntrinsicInterface genericIntrinsicFunction[]{
+    {"abs", {{"a", SameIntOrReal}}, SameIntOrReal},
```

Comment:
It might help document to have some no-op annotation like "a"_arg. I don't think it's unreadable as is though.